File: frontend/src/canvasReferences.ts
import type { CanvasFlowEdge, CanvasFlowNode } from './canvasTypes'
import { slotKindOfHandle } from './canvasRules'

export type ReferenceKind = 'image' | 'video' | 'audio'
export type ReferenceBinding = { edge_id: string; alias: string; description: string }
export type ReferenceRow = ReferenceBinding & { kind: ReferenceKind; name: string; tag: string; source: string }
const names = { image: '图片', video: '视频', audio: '音频' }

export function referenceTag(model: string, kind: ReferenceKind, number: number) {
  if (model.startsWith('happyhorse')) return `[${{ image: 'Image', video: 'Video', audio: 'Audio' }[kind]} ${number}]`
  if (model.startsWith('wan')) return `${{ image: '图', video: '视频', audio: '音频' }[kind]} ${number}`
  return `第 ${number} ${kind === 'image' ? '张参考图片' : kind === 'video' ? '段参考视频' : '段参考音频'}`
}

export function canvasReferences(id: string, data: Record<string, unknown>, nodes: CanvasFlowNode[], edges: CanvasFlowEdge[]) {
  const saved: ReferenceBinding[] = Array.isArray(data.reference_bindings) ? data.reference_bindings : []
  const byId = new Map(nodes.map(node => [node.id, node]))
  const connected = edges.filter(edge => edge.target === id && ['image', 'video', 'audio'].includes(slotKindOfHandle(edge.targetHandle || '') || '') && byId.has(edge.source))
  const rank = new Map(saved.map((item, index) => [item.edge_id, index]))
  const used = new Set(saved.map(item => item.alias))
  const rows: ReferenceRow[] = []
  for (const kind of ['image', 'video', 'audio'] as const) {
    const group = connected.filter(edge => slotKindOfHandle(edge.targetHandle || '') === kind)
      .sort((a, b) => (rank.get(a.id) ?? Infinity) - (rank.get(b.id) ?? Infinity))
    for (const [index, edge] of group.entries()) {
      const existing = saved.find(item => item.edge_id === edge.id)
      let next = 1
      while (used.has(`${names[kind]}${next}`)) next++
      const alias = existing?.alias || `${names[kind]}${next}`
      used.add(alias)
      const source = byId.get(edge.source)!
      rows.push({ edge_id: edge.id, alias, description: existing?.description || '', kind,
        name: String(source.data.name || source.data.label || `${names[kind]}素材`), source: source.id,
        tag: referenceTag(String(data.model || ''), kind, index + 1) })
    }
  }
  // Keep retired aliases reserved, so reconnecting a different asset cannot
  // silently reuse a reference that remains in the user's prompt.
  const bindings = [...rows.map(({ edge_id, alias, description }) => ({ edge_id, alias, description })),
    ...saved.filter(item => !rows.some(row => row.edge_id === item.edge_id))]
  const promptEdge = edges.find(edge => edge.target === id && edge.targetHandle === 'prompt')
  const connectedPrompt = String(byId.get(promptEdge?.source || '')?.data.text || '').trim()
  const inlinePrompt = String(data.inlinePrompt || '').trim()
  // Merely opening or moving a legacy node must not change its frozen inputs.
  const enabled = saved.length > 0 || /\{\{[^{}]+\}\}/.test(connectedPrompt + inlinePrompt)
  const prompt = [connectedPrompt, !promptEdge || enabled ? inlinePrompt : ''].filter(Boolean).join('\n\n')
  return { rows, bindings, enabled, prompt }
}

export function referencePrompt(prompt: string, rows: ReferenceRow[], enabled = true) {
  const missing = new Set<string>()
  const resolved = prompt.replace(/\{\{([^{}]+)\}\}/g, (token, alias: string) => {
    const row = rows.find(item => item.alias === alias.trim())
    if (!row) { missing.add(alias.trim()); return token }
    return row.tag
  })
  const legend = rows.map(row => `${row.tag}：${row.description.trim() || row.name}`)
  return { prompt: resolved + (enabled && legend.length ? `\n\n参考素材对应关系（各类型分别按发送顺序编号）：\n${legend.join('\n')}` : ''), missing: [...missing] }
}

File: backend/app/canvas_references.py
"""Stable canvas reference tokens -> provider prompt ordinals and media order.

Wan 3: https://help.aliyun.com/zh/model-studio/wan3-video-generation-api-reference
HappyHorse: https://help.aliyun.com/zh/model-studio/happyhorse-reference-to-video-api-reference
MiniMax documents typed media order, but no special reference-token syntax;
use explicit natural-language ordinals with the accompanying reference legend.
"""
import re

from fastapi import HTTPException

KINDS = {"image": "图片", "video": "视频", "audio": "音频"}
TOKEN = re.compile(r"\{\{([^{}]+)\}\}")


def reference_tag(model: str, kind: str, number: int) -> str:
    if model.startswith("happyhorse"):
        return f"[{dict(image='Image', video='Video', audio='Audio')[kind]} {number}]"
    if model.startswith("wan"):
        return f"{dict(image='图', video='视频', audio='音频')[kind]} {number}"
    return f"第 {number} {dict(image='张参考图片', video='段参考视频', audio='段参考音频')[kind]}"


def bind_references(model: str, prompt: str, media: list[dict], bindings: list[dict]) -> tuple[str, list[dict]]:
    if not isinstance(bindings, list):
        raise HTTPException(400, "素材引用设置格式不正确")
    by_edge = {}
    used = set()
    for binding in bindings:
        if not isinstance(binding, dict):
            raise HTTPException(400, "素材引用设置格式不正确")
        edge_id = binding.get("edge_id")
        alias = binding.get("alias")
        description = binding.get("description", "")
        if not isinstance(edge_id, str) or not isinstance(alias, str) or not re.fullmatch(r"[\w\-]{1,40}", alias):
            raise HTTPException(400, "素材引用名称或连线无效")
        if edge_id in by_edge or alias in used:
            raise HTTPException(400, "素材引用名称或连线不能重复")
        if not isinstance(description, str) or len(description) > 200:
            raise HTTPException(400, "素材用途不能超过 200 个字符")
        by_edge[edge_id] = binding
        used.add(alias)
    rank = {item["edge_id"]: index for index, item in enumerate(bindings)}
    ordered = []
    tags = {}
    legend = []
    for kind, label in KINDS.items():
        group = sorted((item for item in media if item["kind"] == kind),
                       key=lambda item: rank.get(item.get("_edge_id"), len(bindings)))
        for index, item in enumerate(group, 1):
            binding = by_edge.get(item.get("_edge_id"), {})
            number = 1
            while f"{label}{number}" in used:
                number += 1
            alias = binding.get("alias") or f"{label}{number}"
            used.add(alias)
            tag = reference_tag(model, kind, index)
            tags[alias] = tag
            name = item.get("name") or f"{label}素材"
            legend.append(f"{tag}：{binding.get('description', '').strip() or name}")
            ordered.append({key: value for key, value in item.items() if key != "_edge_id"})

    def replace(match):
        alias = match[1].strip()
        if alias not in tags:
            raise HTTPException(400, f"引用素材「{alias}」已断开或不存在，请重新连接或修改提示词")
        return tags[alias]

    explicit = bool(bindings) or bool(TOKEN.search(prompt))
    prompt = TOKEN.sub(replace, prompt)
    if explicit and legend:
        prompt += "\n\n参考素材对应关系（各类型分别按发送顺序编号）：\n" + "\n".join(legend)
    return prompt, ordered
